src/setSpiderMonkeyException.cc

```diff
@@ -0,0 +1,78 @@
+/**
+ * @file setSpiderMonkeyException.cc
+ * @author Caleb Aikens (caleb@distributive.network)
+ * @brief Call this function whenever a JS_* function call fails in order to set an appropriate python exception (remember to also return NULL)
+ * @version 0.1
+ * @date 2023-02-28
+ *
+ * @copyright Copyright (c) 2023
+ *
+ */
+
+#include "include/modules/pythonmonkey/pythonmonkey.hh"
+#include "include/setSpiderMonkeyException.hh"
+#include "include/StrType.hh"
+
+#include <jsapi.h>
+
+#include <codecvt>
+#include <locale>
+
+#include <Python.h>
+
+void setSpiderMonkeyException(JSContext *cx) {
+  if (!JS_IsExceptionPending(cx)) {
+    PyErr_SetString(SpiderMonkeyError, "Spidermonkey failed, but spidermonkey did not set an exception.");
+    return;
+  }
+  JS::ExceptionStack exceptionStack(cx);
+  if (!JS::GetPendingExceptionStack(cx, &exceptionStack)) {
+    PyErr_SetString(SpiderMonkeyError, "Spidermonkey set an exception, but was unable to retrieve it.");
+    return;
+  }
+  JS::RootedObject exceptionObject(cx);
+  if (!JS_ValueToObject(cx, exceptionStack.exception(), &exceptionObject)) {
+    PyErr_SetString(SpiderMonkeyError, "Spidermonkey set an exception, but the exception could not be converted to an object.");
+    return;
+  }
+
+  /**
+   * the resulting python error string will be in the format:
+   * "Error in file <filename>, on line <lineno>:
+   * <offending line of code if relevant, nothing if not>
+   * <if offending line is present, then a '^' pointing to the relevant token>
+   * <spidermonkey error message>
+   * Stack Track:
+   * <stack trace>"
+   *
+   */
+  std::stringstream outStrStream;
+
+  JSErrorReport *errorReport = JS_ErrorFromException(cx, exceptionObject);
```

Comment:
Maybe refactor to something like https://hg.mozilla.org/releases/mozilla-esr102/file/tip/js/src/shell/js.cpp#l10176

Reply:
`JS::ErrorReportBuilder` can output the error type like `TypeError: `